### DISCUSSION TITLE
Increase bandwidth.

### DISCUSSION THREAD

**Discussion by TroyKomodo** [pion/webrtc#1613] [Q&A]
Hi, I was curious about how to increase the bandwidth on the WebRTC channel for higher quality video

Currently I can send 700Kbps at 60FPS and all 60frames make it to the client, but 700Kbps on 1080p video looks like trash.

if I send 2000Kbps the video plays back at ~ 1FPS.

🚀

**Comment by Sean-Der** [comment]
Hey @TroyDota 

Right now you can use [REMB](https://webrtcglossary.com/remb/) this is a message from Pion to the sender that instructs it to use a certain bitrate you can see that [here](https://github.com/pion/webrtc/blob/master/examples/simulcast/main.go#L103-L105)

I am working on add transport wide congestion control, this means the browser will always send the best bitrate and the developer doesn't need to do anything. I don't have an estimate for this yet.

**Reply by TroyKomodo** [reply]
Hey thanks for the quick response.

I looked at the example 👍 in my case i am distributing audio/video from the server to the client, does the server then control the bandwidth entirely or does the client still have a say in the bandwidth of the stream? In essence is there something I must set client side to allow for the increase in bandwidth request from the server?

**Reply by TroyKomodo** [reply]
Since I am only receiving video and audio on the browser.

**Reply by Sean-Der** [reply]
The client *should* respect what you send, but it doesn’t have to. Chrome and FireFox will work. 

GStreamer and Pion will not (since encoder is decoupled).

**Reply by TroyKomodo** [reply]
Is this after i specify a bandwidth increase or should the client respect any audio/video tracks i push it?

**Comment by TroyKomodo** [comment]
Basically what my end goal is to play back a video from disk but it lags and shutters due to the bandwidth, at 500Kbps playback is fine.

**Comment by TroyKomodo** [comment]
```
instance.cmd = exec.Command("ffmpeg",
  "-i", "in.mp4",
  "-map", "0:v",
  "-c:v", "vp8",
  "-deadline", "realtime",
  "-g", "60",
  "-minrate", "6M",
  "-maxrate", "6M",
  "-b:v", "6M",
  "-cpu-used", "8",
  "-f", "ivf", "out.ivf",
  "-map", "0:a",
  "-c:a", "libopus",
  "-ac", "2",
  "-page_duration", "10000",
  "-f", "ogg", "out.ogg",
)
```

using this command to transcode the audio and video into vp8 and opus the audio plays fine on stream but the video doesnt play at all.

**Reply by TroyKomodo** [reply]
Changing the bitrate to 1M allows the video to play at very low quality and then at 2M the video plays but FPS is 1. Anything higher than 2M and the video doesnt play.

**Reply by TroyKomodo** [reply]
In my actual code i am not writing to disk but rather to a unix socket and then reading from that in golang but it is effectively the same thing.

So my playback from disk example  is slightly different to incorporate  the file source as a socket.

**Reply by TroyKomodo** [reply]
![image](https://user-images.githubusercontent.com/49777269/103151066-535fb080-4783-11eb-8935-b597aab8ca59.png)
The video using the above command is very very very poor quality.

Yet i tested the ffmpeg output by writing it to disk and reading that and it was the quality expected.

**Comment by TroyKomodo** [comment]
```
(*ivf.Frame)(0xc0014c3d10)({
 FileHeader: (*ivf.FileHeader)(0xc0003a2680)({
  signature: (string) (len=4) "DKIF",
  version: (uint16) 0,
  headerSize: (uint16) 32,
  FourCC: (string) (len=4) "VP80",
  Width: (uint16) 1920,
  Height: (uint16) 1080,
  TimebaseDenominator: (uint32) 60,
  TimebaseNumerator: (uint32) 1,
  NumFrames: (uint32) 4294967295,
  unused: (uint32) 4294967295
 }),
 FrameHeader: (*ivf.FrameHeader)(0xc00155b310)({
  FrameSize: (uint32) 67383,
  Timestamp: (uint64) 2130
 }),
 Data: ([]uint8) <nil>
})
2020/12/26 13:50:41 65524 underflow
(*ivf.Frame)(0xc0014c3d70)({
 FileHeader: (*ivf.FileHeader)(0xc0003a2680)({
  signature: (string) (len=4) "DKIF",
  version: (uint16) 0,
  headerSize: (uint16) 32,
  FourCC: (string) (len=4) "VP80",
  Width: (uint16) 1920,
  Height: (uint16) 1080,
  TimebaseDenominator: (uint32) 60,
  TimebaseNumerator: (uint32) 1,
  NumFrames: (uint32) 4294967295,
  unused: (uint32) 4294967295
 }),
 FrameHeader: (*ivf.FrameHeader)(0xc00155b650)({
  FrameSize: (uint32) 1400406469,
  Timestamp: (uint64) 3901413713953030253
 }),
 Data: ([]uint8) <nil>
})
```
As it turns out the socket has some sort of max limit. And the frame size is larger than the number of bytes left in the socket so it  goes to a new iteration of the loop and then reads an invalid frame header on the next run due to the bytes of the pervious frame going over into the next iteration.

**Reply by TroyKomodo** [reply]
Perhaps i convert the socket into a channel and then read from the channel, <- blocking so that i wait till the all the bytes are ready...

**Comment by Sean-Der** [comment]
@TroyDota You are seeing packet loss because of congestion. If you send more then the network can handle WebRTC doesn't cache, it drops data. If you try this same video but over your local network it will work. 

Right now Pion WebRTC doesn't have built-in BWE (to tell you how much bitrate you have available) you can look at the receiver reports and get the percentage loss to generate yourself.

**Reply by TroyKomodo** [reply]
Ic, my entire idea is actually to have ffmpeg transcode to the quality options and then read them in via socket ( which now works it was the max 16bit int problem) and then to calculate packet loss on the client side and have the client dynamically request (via data channel) the quality in which they want to receive (or if the  client wants they can force a quality setting and then we will send that to them), kinda like HLS ABR  except WebRTC.

**Comment by TroyKomodo** [comment]
Also another unrelated note to this particular discussion, do you recommend using webrtc for pushing video around on servers (there can be no packet loss on these pushes)? Or should i resort to standard websocket? All servers have 1Gigabit upstream and some have 10Gigabit.

**Comment by EricOreg** [comment]
@TroyKomodo Have you found the solution for your problem (WebRTC doesn't support bit rate over 2 mbps) yet? I am having the exact same problem now. Even with the direct connection between client laptop and the server with static IP address, the bottom part of the video still shows blocks like the image you uploaded before. As you have verified, encoder, decoder and network bandwidth are not the cause of the problem. Seems like something in the PION server or Webrtc caused the video problem.

**Reply by TroyKomodo** [reply]
I never found a solution, i stopped trying.
